### PR TITLE
parser: implement Restore for PrepareStmt ExplainStmt ExecuteStmt Dea…

### PR DESCRIPTION
**Pull request by eiantee**
…llocateStmt

### CODE REVIEW

File: ast/misc_test.go [hunk 1]
```diff
@@ -203,3 +203,50 @@ func (ts *testMiscSuite) TestTableOptimizerHintRestore(c *C) {
 	}
 	RunNodeRestoreTest(c, testCases, "select /*+ %s */ * from t1 join t2", extractNodeFunc)
 }
+
+func (ts *testMiscSuite) TestExplainStmtRestore(c *C) {
```
Comment (leoppro) [inline ast/misc_test.go]: test `StmtNode` in `parser_test.go`

File: ast/misc.go [hunk 2]
```diff
@@ -228,7 +277,20 @@ type ExecuteStmt struct {
...
+	ctx.WriteKeyWord("EXECUTE ")
+	ctx.WriteName(n.Name)
+	if n.UsingVars != nil && len(n.UsingVars) > 0 {
+		ctx.WriteKeyWord(" USING ")
+		for i, val := range n.UsingVars {
+			if i != 0 {
+				ctx.WritePlain(",")
+			}
+			if err := val.Restore(ctx); err != nil {
+				return errors.Annotate(err, fmt.Sprintf("An error occurred while restore ExecuteStmt.UsingVars index %d", i))
```
Comment (leoppro) [inline ast/misc.go]:
```suggestion
				return errors.Annotatef(err, "An error occurred while restore ExecuteStmt.UsingVars index %d", i)
```

Comment (eiantee) [reply ast/misc.go]: thank you for suggestion，it has been fixed！

File: ast/misc.go [hunk 3]
```diff
@@ -134,7 +134,34 @@ type ExplainStmt struct {
...
+				return errors.Annotate(err, "An error occurred while restore ExplainStmt.ShowStmt.Column")
+			}
+		}
+		return nil
+	}
+	ctx.WriteKeyWord("EXPLAIN ")
+	if n.Analyze {
+		ctx.WriteKeyWord("ANALYZE ")
+	} else {
+		if strings.ToLower(n.Format) == ExplainFormatDOT {
```
Comment (kennytm) [inline ast/misc.go]: Change the condition to

```suggestion
		if strings.ToLower(n.Format) != ExplainFormatROW {
```

or just let `Restore()` always write out the FORMAT. MySQL supports two entirely different formats (`JSON` and `TRADITIONAL`) and I don't think there's any reason we cannot expand the list of formats in the future.

File: ast/misc.go [hunk 4]
```diff
@@ -165,7 +192,23 @@ type PrepareStmt struct {
 
 // Restore implements Node interface.
 func (n *PrepareStmt) Restore(ctx *RestoreCtx) error {
-	return errors.New("Not implemented")
+	ctx.WriteKeyWord("PREPARE ")
+	if n.Name == "" {
+		return errors.New("An error occurred while restore PrepareStmt.Name")
```
Comment (kennytm) [inline ast/misc.go]: What's wrong with an empty name? The following is valid on MySQL 8:

```sql
PREPARE `` FROM 'select 1';
EXECUTE ``;
```

File: ast/misc.go [hunk 5]
```diff
@@ -195,7 +238,13 @@ type DeallocateStmt struct {
 
 // Restore implements Node interface.
 func (n *DeallocateStmt) Restore(ctx *RestoreCtx) error {
-	return errors.New("Not implemented")
+	if n.Name == "" {
+		return errors.New("An error occurred while restore DeallocateStmt.Name")
+	}
+	ctx.WriteKeyWord("DEALLOCATE ")
+	ctx.WriteKeyWord("PREPARE ")
```
Comment (kennytm) [inline ast/misc.go]: Combine these two statements.

```go
ctx.WriteKeyWord("DEALLOCATE PREPARE ");
```

File: ast/misc_test.go (hunk 1 above)
Comment (leoppro) [inline ast/misc_test.go]: i think we can remove tests in `misc_test.go`

**Comment by eiantee** [comment]
I have modified those as you suggested and added some empty-name test case

File: ast/misc.go [hunk 6]
```diff
@@ -228,7 +268,20 @@ type ExecuteStmt struct {
 
 // Restore implements Node interface.
 func (n *ExecuteStmt) Restore(ctx *RestoreCtx) error {
-	return errors.New("Not implemented")
+	ctx.WriteKeyWord("EXECUTE ")
+	ctx.WriteName(n.Name)
+	if n.UsingVars != nil && len(n.UsingVars) > 0 {
```
Comment (kennytm) [inline ast/misc.go]:
```suggestion
	if len(n.UsingVars) > 0 {
```

It is unnecessary to check `n.UsingVars != nil` since `len(nil) == 0`.

**Pull request merged**